src/client/Client.cc

```diff
@@ -7180,11 +8003,15 @@ int Client::rename(const char *relfrom, const char *relto, const UserPerm& perm,
   if (r < 0)
     return r;
 
-  std::scoped_lock lock(client_lock);
   if (cct->_conf->client_permissions) {
-    int r = may_delete(fromdir.get(), fromname.c_str(), perm);
-    if (r < 0)
-      return r;
+    {
+      std::scoped_lock il(fromdir->inode_lock);
+      int r = may_delete(fromdir.get(), fromname.c_str(), perm);
+      if (r < 0)
+        return r;
+    }
+
+    std::scoped_lock il(todir->inode_lock);
```

Comment:
Wait, so you don't need to hold the locks for both fromdir and todir during a cross-directory rename? Why not?

If it turns out that you do need to hold both locks here, then you have an issue. You really can't rely on parent/child relationships to determine the ordering since the fromdir and todir may not be nested.

FWIW, the kernel does this by looking at the value of the dentry pointer and the lower one goes first. You may need to do something similar here instead of trying to rely on parent/child ordering for the locking.

Reply:
Yeah, this need to be improved, I am still thinking how to make it work well if `todir` and `fromdir` are not be nested.

Currently there no case will the inode locks will nest unless they have the parent/child relation.

Checked the kernel code, it seems it must hold the global `s_vfs_rename_mutex` before there has inode locks nesting to avoid the dead lock. This sounds a good idea.

Reply:
Yeah, it's not terribly scalable (i.e., it probably ought to be a per-sb mutex), but that's what we have in the kernel for now.

Reply:
Yeah, locally the `rename lock` patch is ready and can reduce many lock thrashing :-)

Reply:
The logic is similar:
1, if `fromdir` equals to `todir`, it's simple, just take the `fromdir`'s inode lock

For case1, there is no need to take the global `s_vfs_rename_mutex`.

2, if `fromdir` is ancester, then the lock order will be `fromdir->inode_lock` --> `todir->inode_lock`
3, if `todir` is ancester, then the lock order will be `todir->inode_lock` --> `fromdir->inode_lock`
4, or take the `inode_locks` in any order will be okay.

For case2~4, it will always take the global `s_vfs_rename_mutex` first.

Here the case2 and case3 are necessary and will always keep the lock order if they have parent/child relationship. Then it could concurrent run with other threads where the inode locks may nest, and won't cause the dead lock.